tripy/tripy/frontend/trace/ops/base.py

```diff
@@ -73,80 +72,27 @@ def build(cls, inputs: List["Tensor"], *args, num_outputs=1, **kwargs) -> Union[
         of returning a list of output tensors.
         """
 
-        from tripy.frontend.shape import Shape, ShapeScalar
         from tripy.frontend.tensor import Tensor
+        from tripy.frontend.dimension_size import DimensionSize
+
+        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
+        # For any mixed operations, DimensionSize must be casted up to Tensor.
+        TensorType = DimensionSize if all(isinstance(inp, DimensionSize) for inp in inputs) else Tensor
```

Comment:
Still not completely sure of how to implement the logic to determine whether the output should be a DimensionSize or a Tensor. There are two parts to this:
1. What should the decorator convert inputs to? If it always makes Tensors, the operations will never see DimensionSizes and will therefore never be able to indicate the output type accurately. 
2. What type should the op generate?

Reply:
Shall we only support a few ops to return `DimensionSize`? E.g. only binary and unary elementwise. I did not see other use cases. To return `DimensionSize`, the op must follow quite strict rules, e.g. I/O rank must be 0, dtype must be int32, it may be easier to just white-list the ops and let them override some function to provide the type information.

Reply:
I agree. I'm stuck on the mechanism for whitelisting though. What we would need is for the frontend APIs to somehow have knowledge of the original input types before they are converted to tensors. With that in place, we could just have `convert_to_tensors` always convert to the `Tensor` base class - since `Tensor`/`DimensionSize` are just frontend constructs, we can actually consume `Tensor`s and produce `DimensionSize`. 

One approach is to attach the original type as a private field to the tensor produced by `convert_to_tensors`:
```py
@convert_to_tensors()
def func(a: TensorLike, b: TensorLike):
    return_type = Tensor
    if a._original_type is DimensionSize and b._original_type is DimensionSize:
        return_type = DimensionSize
```
But I'm wondering if there is a better/cleaner approach than this.

Reply:
Is there any reason not to convert to `DimensionSize` if the criteria are met?

Reply:
Since `DimensionSize` -> `Tensor` is a one-way conversion, we should try to keep things as `DimensionSize`s where possible. The problem is that the logic that I've currently implemented isn't completely right - not all Python integers actually represent dimension sizes and not all ops produce dimension sizes even if all their inputs are dimension sizes.

Reply:
Is it possible to wrap another decorator to decide the return type? e.g.

```py
@infer_return_type
@convert_to_tensors()
def func(...)
```

Reply:
> not all Python integers actually represent dimension sizes

We could present `DimensionSize` as meaning that the value _could_ be used to represent a dimension size, but maybe that's too clever by half. I think using a method or perhaps another decorator to indicate a policy for wrapping might be a sensible approach if isn't a way to avoid the issue.

Reply:
Yes, an outer decorator is a good idea. Will update the PR accordingly.

Reply:
Updated this as we discussed so that we always try casting to DimensionSize where possible.